carball/analysis/kickoff_detection/kickoff_analysis.py

```diff
@@ -0,0 +1,159 @@
+import logging
+import time
+from typing import Dict, List, Callable
+from bisect import bisect_left
+import numpy as np
+import pandas as pd
+
+from carball.analysis.constants.basic_math import position_column_names
+from ...generated.api import game_pb2
+from ...generated.api.player_pb2 import Player
+from ...generated.api.stats.events_pb2 import Hit
+from ...generated.api.stats.kickoff_pb2 import KickoffStats
+from ...generated.api.stats import kickoff_pb2 as kickoff
+from ...json_parser.game import Game
+from ...analysis.hit_detection.base_hit import BaseHit
+from ...analysis.simulator.ball_simulator import BallSimulator
+from ...analysis.simulator.map_constants import *
+
+
+logger = logging.getLogger(__name__)
+
+
+class BaseKickoff:
+
+    def is_kickoff(data_frame, index):
+        ball_position = data_frame.ball.loc[index, position_column_names]
+        if ((ball_position.pos_x == 0.0) and
+            (ball_position.pos_y == 0.0)): #z=92.74
+          return True
+        else:
+          return False
+
+    @staticmethod
+    def get_kickoffs_from_game(game: Game, proto_game: game_pb2, id_creation:Callable,
+                               player_map: Dict[str, Player],
+                               data_frame: pd.DataFrame, kickoff_frames: pd.DataFrame) -> Dict[int, KickoffStats]:
+        kickoffs = dict()
+        for frame in kickoff_frames:
+            cur_kickoff = proto_game.kickoff_stats.add()
+            end_frame = frame
+            while BaseKickoff.is_kickoff(data_frame, end_frame):
```

Comment:
There may be a way to do this where we can use pandas to search for the first frame where it is no longer a kickoff without needing to do this.

But I'm not sure without spending time looking into it

Reply:
It's already a duplicate of work being done in get_kickoff_frames:
```python
kickoff_frames, first_touch_frames = self.get_kickoff_frames(self.game, self.protobuf_game, data_frame)
```
I just copied what my one-off stuff was doing instead of finding a way to shuffle that around